plugins/DISTRHO-Ports-master/libs/juced/source/dependancies/dssi-1.0.0/doc/dssi_osc_send.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH dssi_osc_send 1 "January 5th, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
dssi_osc_send \- send OSC messages to DSSI applications.
.SH SYNOPSIS
.B dssi_osc_send
.I <option> <OSC URL> <values>
.SH DESCRIPTION
.B dssi_osc_send
sends DSSI OSC methods to DSSI applications.
.B dssi_osc_send
is typically used to control DSSI hosts (possibly for scripted
operation), or to test DSSI user interfaces.
.SH OPTIONS
.TP
.B -c
Send a `control' message.  Requires two values to be supplied
in the <values> section: control port and value.
For example, to set control port 1 to 0.1:
  dssi_osc_send -c <OSC URL> 1 0.1
.TP
.B -p
Send a `program' message.  Requires two values to be supplied
in the <values> section: bank and program number.
For example, to select bank 0 program number 7:
  dssi_osc_send -p <OSC URL> 0 7
.TP
.B -m
Send `midi' message.  Requires four numbers to be supplied
in the <values> section.
For example, to  send a note on, middle C, velocity 64:
  dssi_osc_send -m <OSC URL> 0 144 60 64
.TP
.B -C
Send a `configure' message. The <values> are the key and
value pair. For example, to send key `load' and value
`/tmp/patches.pat':
  dssi_osc_send -C <OSC URL> load /tmp/patches.pat
.TP
.B -s
Send a `show' message.
  dssi_osc_send -s <OSC URL>
.TP
.B -h
Send a `hide' message.
  dssi_osc_send -h <OSC URL>
.TP
.B -q
Send a `quit' message.
  dssi_osc_send -q <OSC URL>
.TP
.B -e
Send a `exiting' message.
  dssi_osc_send -e <OSC URL>
.TP
.B <OSC URL>
The URL to connect to.  For example:
  osc.udp://localhost:19383/dssi/test.1
  osc.udp://my.host.org:10886/dssi/xsynth-dssi/Xsynth/chan00
.SH SEE ALSO
.BR dssi_osc_update (1).
.br
.SH AUTHOR
This manual page was originally created by Mark Hymers from the help
text of the application, for the Debian project (but may be freely
used by others).


plugins/DISTRHO-Ports-master/libs/juced/source/dependancies/dssi-1.0.0/examples/dssi_osc_send.c
/*
 *  This program is in the public domain
 *
 *  $Id: dssi_osc_send.c,v 1.5 2004/07/04 22:24:51 smbolton Exp $
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <lo/lo.h>

void
usage(char *program_name)
{
    char *base_name = strrchr(program_name, '/');
    
    if (base_name && *(base_name + 1) != 0) {
        base_name += 1;
    } else {
        base_name = program_name;
    }
    fprintf(stderr, "usage: %s <option> <OSC URL> <values>\n\n", program_name);
    fprintf(stderr, "example OSC URLs:\n\n"
                    "  osc.udp://localhost:19383/dssi/test.1\n"
                    "  osc.udp://my.host.org:10886/dssi/xsynth-dssi/Xsynth/chan00\n\n");
    fprintf(stderr, "example 'control' message (set control port 1 to 0.1):\n\n"
                    "  %s -c <OSC URL> 1 0.1\n\n", base_name);
    fprintf(stderr, "example 'program' message (select bank 0 program number 7):\n\n"
                    "  %s -p <OSC URL> 0 7\n\n", base_name);
    fprintf(stderr, "example 'midi' message (send a note on, middle C, velocity 64):\n\n"
                    "  %s -m <OSC URL> 0 144 60 64\n\n", base_name);
    fprintf(stderr, "example 'configure' message (send key 'load' and value '/tmp/patches.pat'):\n\n"
                    "  %s -C <OSC URL> load /tmp/patches.pat\n\n", base_name);
    fprintf(stderr, "messages requiring no additional arguments:\n\n"
                    "  %s -s <OSC URL>      (send 'show' message)\n", base_name);
    fprintf(stderr, "  %s -h <OSC URL>      (send 'hide' message)\n", base_name);
    fprintf(stderr, "  %s -q <OSC URL>      (send 'quit' message)\n", base_name);
    fprintf(stderr, "  %s -e <OSC URL>      (send 'exiting' message)\n", base_name);
    exit(1);
}

int main(int argc, char *argv[])
{
    lo_address a;
    char *url, *host, *port, *path, *method;
    char mode = '-';
    char full_path[256];

    int ladspa_port, int0, int1;
    float value;
    unsigned char midi[4];

    if (argc < 3) {
        usage(argv[0]);
        /* does not return */
    }
    url = argv[1];
    if (argv[1][0] == '-') {
        if (argv[1][1] == 'c' ||
            argv[1][1] == 'p' ||
            argv[1][1] == 'm' ||
            argv[1][1] == 'C' ||
            argv[1][1] == 's' ||
            argv[1][1] == 'h' ||
            argv[1][1] == 'q' ||
            argv[1][1] == 'e') {
            mode = argv[1][1];
            url = argv[2];
        } else {
            usage(argv[0]);
        }
    }

    host = lo_url_get_hostname(url);
    port = lo_url_get_port(url);
    path = lo_url_get_path(url);
    a = lo_address_new(host, port);

    switch (mode) {

      case '-': /* old 'control' syntax */
        if (argc != 4) usage(argv[0]);
        ladspa_port = atoi(argv[2]);
        value = atof(argv[3]);
        printf("sending osc.udp://%s:%s%s %d %f\n", host, port, path,
               ladspa_port, value);
        lo_send(a, path, "if", ladspa_port, value);
        break;

      case 'c': /* new 'control' syntax */
        if (argc != 5) usage(argv[0]);
        snprintf(full_path, 255, "%s/control", path);
        ladspa_port = atoi(argv[3]);
        value = atof(argv[4]);
        printf("sending osc.udp://%s:%s%s %d %f\n", host, port, full_path,
               ladspa_port, value);
        lo_send(a, full_path, "if", ladspa_port, value);
        break;

      case 'p': /* send 'program' message */
        if (argc != 5) usage(argv[0]);
        snprintf(full_path, 255, "%s/program", path);
        int0 = atoi(argv[3]);
        int1 = atoi(argv[4]);
        printf("sending osc.udp://%s:%s%s %d %d\n", host, port, full_path,
               int0, int1);
        lo_send(a, full_path, "ii", int0, int1);
        break;

      case 'm': /* send 'midi' message */
        if (argc < 5 || argc > 7) usage(argv[0]);
        snprintf(full_path, 255, "%s/midi", path);
        midi[0] = atoi(argv[3]);                 /* "port id", according to OSC spec */
        midi[1] = atoi(argv[4]);                 /* status byte */
        midi[2] = argc > 5 ? atoi(argv[5]) : 0;  /* data byte 1 */
        midi[3] = argc > 6 ? atoi(argv[6]) : 0;  /* data byte 2 */
        printf("sending osc.udp://%s:%s%s [%02x %02x %02x %02x]\n", host, port, full_path,
               midi[0], midi[1], midi[2], midi[3]);
        lo_send(a, full_path, "m", midi);
        break;

      case 'C': /* send 'configure' message */
        if (argc != 5) usage(argv[0]);
        snprintf(full_path, 255, "%s/configure", path);
        printf("sending osc.udp://%s:%s%s \"%s\" \"%s\"\n", host, port, full_path,
               argv[3], argv[4]);
        lo_send(a, full_path, "ss", argv[3], argv[4]);
        break;

      case 's': /* send 'show' message */
      case 'h': /* send 'hide' message */
      case 'q': /* send 'quit' message */
      case 'e': /* send 'exiting' message */
        if (argc != 3) usage(argv[0]);
        switch (mode) {
          default:
          case 's':  method = "show";     break;
          case 'h':  method = "hide";     break;
          case 'q':  method = "quit";     break;
          case 'e':  method = "exiting";  break;
        }
        snprintf(full_path, 255, "%s/%s", path, method);
        printf("sending osc.udp://%s:%s%s\n", host, port, full_path);
        lo_send(a, full_path, "");
        break;
    }

    if (lo_address_errno(a)) {
	printf("liblo error: %s\n", lo_address_errstr(a));
    }
    free(host);
    free(port);
    free(path);

    return 0;
}

/* vi:set ts=8 sts=4 sw=4: */
